Fully implement all InsertBefore/InsertAfter overloads

BruceForstall opened the pull request:
Fixes problem where they wouldn't work with insertionPoint==nullptr.

BruceForstall committed:
Fully implement all InsertBefore/InsertAfter overloads

Fixes problem where they wouldn't work with insertionPoint==nullptr.

BruceForstall commented:
@pgavlin PTAL

pgavlin commented on src/jit/lir.cpp:
```diff
@@ -709,11 +665,102 @@ void LIR::Range::InsertBefore(GenTree* insertionPoint, GenTree* node1, GenTree*
+
+    node4->gtPrev = node3;
+    node3->gtPrev = node2;
+    node2->gtPrev = node1;
```
I'm finding the batched gtNext/gtPrev updates a bit hard to read. Can we simply pair the updates for each  node? i.e.

```
node1->gtNext = node2;

node2->gtPrev = node1;
node2->gtNext = node3;

node3->gtPrev = node2;
node3->gtNext = node4;

node4->gtPrev = node3;
```

BruceForstall replied:
I could do that if you find that easier.

pgavlin replied:
That would be great; thanks!

mikedn replied:
Meh, so much duplication :smile:

``` C++
namespace InsertUtils {
    static void Link(GenTree* first, GenTree* second) {
        first->gtNext = second;
        second->gtPrev = first;
    }

    static GenTree* Chain(GenTree* last) {
        return last;
    }

    template<typename ...TRest>
    static GenTree* Chain(GenTree* first, GenTree* second, TRest... rest) {
        Link(first, second);
        return Chain(second, rest...);
    }

    template<typename ...TRest>
    static GenTree* First(GenTree* first, TRest... rest) {
        return first;
    }

    static void InsertChain(GenTree* after, GenTree* first, GenTree* last, GenTree *before) {
        if (after != nullptr) {
            Link(after, first);
        }

        if (before != nullptr) {
            Link(last, before);
        }
    }
};

template<typename ...TNodes>
void InsertAfter(GenTree* after, TNodes... nodes) {
    GenTree* first = InsertUtils::First(nodes...);
    GenTree* last = InsertUtils::Chain(nodes...);
    GenTree* before = after->gtNext;
    InsertUtils::InsertChain(after, first, last, before);
}

template<typename ...TNodes>
void InsertBefore(GenTree* before, TNodes... nodes) {
    GenTree* first = InsertUtils::First(nodes...);
    GenTree* last = InsertUtils::Chain(nodes...);
    GenTree* after = before->gtPrev;
    InsertUtils::InsertChain(after, first, last, before);
}
```

pgavlin commented:
LGTM

BruceForstall committed:
More changes to InsertAfter()/InsertBefore() functions.

1. rearrange node linking per code review feedback.
2. implement range insertion versions using new, common Finish*() helpers.

pgavlin commented:
@BruceForstall: the variadic template magic above is what I was wondering about the other day. Thanks @mikedn for figuring that out so I don't have to :D

BruceForstall commented:
@pgavlin updated PR. Changed range-based insert functions to also use common helper code.

pgavlin commented:
Awesome!

Also, @mikedn: if you want to put together a PR to the LinearBackendIR branch that Bruce and I have been working in with the changes you suggest, we'd be happy to have it.

BruceForstall commented:
I'm not sure that meets my "easy to read/understand" bar, but it is cool...

mikedn commented:
@pgavlin I could do that but I think it's best to do such tweaks after LIR changes reach master, that PR is gigantic as it is...

pgavlin commented:
Fair enough :)
